### PR TITLE
Use collision detection for AI vehicle behaviour

### PR DESCRIPTION
**Pull request by danhedron**
- Provide a tested API for performing area checks against physical objects.

This functionality can be re-used for weapon area check.

Future work
- Document AI vehicle behaviour.

### CODE REVIEW

File: rwengine/src/ai/CharacterController.cpp [hunk 1]
```diff
@@ -384,12 +323,16 @@ bool Activities::DriveTo::update(CharacterObject *character,
     }
 
     // Check whether a pedestrian or vehicle is in our way
-    if (controller->checkForObstacles() == true) {
+    if (controller->checkForObstacles()) {
         currentSpeed = 0.f;
     }
 
+    if (currentSpeed == 0.f) {
```
Comment (ShFil119) [inline rwengine/src/ai/CharacterController.cpp]: `std::fabs(currentSpeed) < 0.1f`

File: rwengine/src/ai/CharacterController.cpp [hunk 2]
```diff
@@ -202,79 +202,18 @@ bool CharacterController::checkForObstacles()
...
-                    } else {
-                        return true;
-                    }
-                }
-            }
+    auto [center, halfSize] = vehicle->obstacleCheckVolume();
+    const auto rotation = vehicle->getRotation();
+    center = vehicle->getPosition() + rotation * center;
+    auto results = test.boxTest(center, halfSize, vehicle->getRotation());
+    for (const auto hit : results) {
```
Comment (ShFil119) [inline rwengine/src/ai/CharacterController.cpp]: Maybe using stl's `std::any_of`?

File: rwengine/src/dynamics/HitTest.cpp [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+    return HitTestWorld(world, ghost);
+}
+
+} // namespace
+
+HitTest::HitTest(btDiscreteDynamicsWorld &world)
+        : _world(world)
+{}
+
+HitTest::~HitTest() {
```
Comment (ShFil119) [inline rwengine/src/dynamics/HitTest.cpp]: `= default`

File: rwengine/src/objects/VehicleObject.cpp [hunk 4]
```diff
@@ -26,6 +26,7 @@
 #include "engine/GameWorld.hpp"
 #include "objects/CharacterObject.hpp"
 
+
```
Comment (ShFil119) [inline rwengine/src/objects/VehicleObject.cpp]: unneeded line

File: rwengine/src/dynamics/HitTest.cpp [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+#pragma warning(default : 4305 5033)
+#endif
+
+namespace {
+
+HitTest::TestResult HitTestWorld(btDiscreteDynamicsWorld& world, btPairCachingGhostObject& tester)
+{
+    world.addCollisionObject(&tester);
+
+    HitTest::TestResult result;
```
Comment (madebr) [inline rwengine/src/dynamics/HitTest.cpp]: Maybe reserve space in the `TestResult` object?
```suggestion
    HitTest::TestResult result;
    result.reserve(tester.getNumOverlappingObjects())
```

File: rwengine/src/dynamics/HitTest.hpp [hunk 6]
```diff
@@ -0,0 +1,37 @@
+#ifndef _RWENGINE_HITTEST_HPP_
+#define _RWENGINE_HITTEST_HPP_
+class GameObject;
```
Comment (madebr) [inline rwengine/src/dynamics/HitTest.hpp]: nit: put all forward declared classes/structs in one place after the includes.

File: tests/test_HitTest.cpp [hunk 7]
```diff
@@ -0,0 +1,94 @@
+#include <boost/test/unit_test.hpp>
+#include "test_Globals.hpp"
+#include <dynamics/HitTest.hpp>
+#include <engine/GameWorld.hpp>
+#include <dynamics/CollisionInstance.hpp>
+#ifdef _MSC_VER
+#pragma warning(disable : 4305 5033)
```
Comment (madebr) [inline tests/test_HitTest.cpp]: note: after applying #678 , 5033 can be removed

File: tests/test_HitTest.cpp [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+#endif
+#include <btBulletDynamicsCommon.h>
+#include <BulletCollision/CollisionDispatch/btGhostObject.h>
+#ifdef _MSC_VER
+#pragma warning(default : 4305 5033)
+#endif
+
+using std::make_unique;
```
Comment (madebr) [inline tests/test_HitTest.cpp]: `make_unique` and `unique_ptr` are used once in this file. So these lines cause more text :wink:

File: rwgame/RWGame.cpp [hunk 9]
```diff
@@ -774,15 +775,53 @@ void RWGame::renderDebugPaths(float time) {
...
-            const glm::vec3 pos2 = v->controller->targetNode->position;
+        if (auto vehicle = v->getCurrentVehicle(); vehicle)
+        {
+            if (v->controller->targetNode) {
+                const glm::vec3 pos1 = v->getPosition();
+                const glm::vec3 pos2 = v->controller->targetNode->position;
 
-            btVector3 position1(pos1.x, pos1.y, pos1.z);
-            btVector3 position2(pos2.x, pos2.y, pos2.z);
+                btVector3 position1(pos1.x, pos1.y, pos1.z);
```
Comment (madebr) [inline rwgame/RWGame.cpp]: Can't you use `drawLine(const glm::vec3 &from, const glm::vec3 &to, const glm::vec3 &color)` here?

File: rwgame/RWGame.cpp [hunk 10]
```diff
@@ -775,17 +776,49 @@ void RWGame::renderDebugPaths(float time) {
...
+            };
+            const auto iRotation = (vehicle->getRotation());
+            const auto rCenter = iRotation * center;
+            std::transform(corners.begin(), corners.end(), corners.begin(),
+                           [&](const auto& p) -> glm::vec3 {
+                               return vehicle->getPosition() + rCenter + iRotation * p;
+                           });
+
+            static const glm::vec3 color2(1.f, 0.f, 0.f);
+            debug.drawLine(corners[0], corners[1], color2);
```
Comment (ShFil119) [inline rwgame/RWGame.cpp]: nit: maybe wrapping it with method `drawBox`?

**Pull request merged**